### PR TITLE
feature/google-workspace-ldap

### PR DESCRIPTION
**Pull request by jrouzierinverse**
Description
===========
Add support for the Google Workspace LDAP as a source.

NEWS file entries
=======================

Enhancements
-------------
* LDAP Source now supports client certificates
* New Source for Google Workspace LDAP

Delete branch after merge
=========================
NO

Checklist
=========
- [ ] Document the feature
- [x] Add unit tests
- [ ] Add acceptance tests (TestLink)

### CODE REVIEW

File: html/pfappserver/lib/pfappserver/Form/Config/Source/GoogleWorkspaceLDAP.pm [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+use strict;
+use warnings;
+use HTML::FormHandler::Moose;
+use pf::Authentication::Source::GoogleWorkspaceLDAPSource;
+extends 'pfappserver::Form::Config::Source::LDAP';
+our $META = pf::Authentication::Source::GoogleWorkspaceLDAPSource->meta;
+
+
+has_field client_certificate => (
```
Comment (julsemaan) [inline html/pfappserver/lib/pfappserver/Form/Config/Source/GoogleWorkspaceLDAP.pm]: What's the difference between client_certificate and client_cert, it seems pretty confusing to have these 2 fields named that way

File: html/pfappserver/root/src/views/Configuration/sources/_components/FormTypeLdap.vue [hunk 2]
```diff
@@ -95,6 +95,27 @@
       :text="$i18n.t('Realms that will be associated with this source.')"
     />
 
+    <form-group-verify namespace="verify"
+      :column-label="$i18n.t('Verify')"
+      :text="$i18n.t('SSL Verify.')"
+    />
+
+    <form-group-client-cert namespace="client_cert"
+      :column-label="$i18n.t('Client Certificate')"
```
Comment (julsemaan) [inline html/pfappserver/root/src/views/Configuration/sources/_components/FormTypeLdap.vue]: Add a descriptive help here
It should state that this is only necessary when using certificate authentication against the LDAP directory

File: html/pfappserver/root/src/views/Configuration/sources/_components/FormTypeLdap.vue [hunk 3]
```diff
@@ -95,6 +95,27 @@
...
+    />
+
+    <form-group-client-key namespace="client_key"
+      :column-label="$i18n.t('Client Key')"
```
Comment (julsemaan) [inline html/pfappserver/root/src/views/Configuration/sources/_components/FormTypeLdap.vue]: Add a descriptive help here
It should state that this is only necessary when using certificate authentication against the LDAP directory

File: html/pfappserver/root/src/views/Configuration/sources/_components/FormTypeLdap.vue [hunk 4]
```diff
@@ -95,6 +95,27 @@
...
+    />
+
+    <form-group-ca-path namespace="ca_path"
+      :column-label="$i18n.t('CA Directory')"
```
Comment (julsemaan) [inline html/pfappserver/root/src/views/Configuration/sources/_components/FormTypeLdap.vue]: Add a descriptive help here
It should state that this is only necessary when using certificate authentication against the LDAP directory

File: html/pfappserver/root/src/views/Configuration/sources/_components/FormTypeLdap.vue [hunk 5]
```diff
@@ -95,6 +95,27 @@
...
+    />
+
+    <form-group-ca-file namespace="ca_file"
+      :column-label="$i18n.t('CA File')"
```
Comment (julsemaan) [inline html/pfappserver/root/src/views/Configuration/sources/_components/FormTypeLdap.vue]: Add a descriptive help here
It should state that this is only necessary when using certificate authentication against the LDAP directory

Comment (julsemaan) [reply html/pfappserver/root/src/views/Configuration/sources/_components/FormTypeLdap.vue]: Also, why a CA file and CA directory field, wouldn't just one be enough ?

File: lib/pf/Authentication/Source/GoogleWorkspaceLDAPSource.pm [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+has '+type' => ( default => 'GoogleWorkspaceLDAP' );
+has 'client_certificate' => ( required => 1, default => 'GoogleWorkspaceLDAP' , is => 'rw');
+has '+host' => ( default => sub { ["ldap.google.com"] });
+has '+port' => (default => 636);
+has '+encryption' => ('default' => 'ssl');
+
+=head2 ldap_attributes
+
+Add ldap search attributes for Active Directory
```
Comment (julsemaan) [inline lib/pf/Authentication/Source/GoogleWorkspaceLDAPSource.pm]: That help is not accurate

**Comment by julsemaan** [comment]
Also needed:

- [ ] Documentation
- [ ] Rebase
- [x] @satkunas' blessing for the admin code

File: html/pfappserver/root/src/views/Configuration/sources/_components/TheForm.vue [hunk 7]
```diff
@@ -114,6 +114,7 @@ export const setup = (props) => {
       case 'Facebook':      return FormTypeFacebook //break
       case 'Github':        return FormTypeGithub //break
       case 'Google':        return FormTypeGoogle //break
+      case 'GoogleWorkspaceLDAP':    return FormTypeLdap //break
```
Comment (satkunas) [inline html/pfappserver/root/src/views/Configuration/sources/_components/TheForm.vue]: Adjust tab-stop of previous and next items so they align (prettier).

**Comment by satkunas** [comment]
@julsemaan UI portion looks good. Only a single cosmetic adjustment is needed.

File: html/pfappserver/root/src/views/Configuration/sources/_components/FormTypeLdap.vue (hunk 5 above)
Comment (jrouzierinverse) [reply html/pfappserver/root/src/views/Configuration/sources/_components/FormTypeLdap.vue]: The more flexible would be the directory.
So I will stick to that one.

**Comment by extrafu** [comment]
@jrouzierinverse and @julsemaan What's the status on this PR?

**Comment by julsemaan** [comment]
> @jrouzierinverse and @julsemaan What's the status on this PR?

We need the doc on how to perform the setup and @jrouzierinverse also had to do tweaks on the CA portion of this

Also, we need the certificate fields to be separated from the main LDAP fields (in the team meeting, we spoke about putting this in a separate tab in the source)

**Pull request merged**